File: Problems/Algorithms/142. Linked List Cycle II/linked_list_cycle_II.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func detectCycle(head *ListNode) *ListNode {
    if head == nil {
        return nil
    }

    visited := make(map[*ListNode]bool)
    curr := head
    for curr != nil {
        if _, ok := visited[curr]; ok {
            return curr
        }

        visited[curr] = true
        curr = curr.Next
    }

    return nil
}

File: Problems/Algorithms/142. Linked List Cycle II/linked_list_cycle_II.py
#!/usr/bin/env python3

from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None:
            return None
    
        visited = {}
        curr = head
        while curr:
            if curr in visited:
                return curr
        
            visited[curr] = True
            curr = curr.next
    
        return None
